lib/web_catalog/api_extractor.es6.js

```diff
@@ -140,511 +242,428 @@ foam.CLASS({
-          // is not considered as an interface even extra properties exists.
-          // Eg. chrome.webstore, chrome.runtime.
-
-          // Skip if this function object is Function.prototype;
-          if (id === this.functionProtoId) return;
-          let prototypePropertyId = og.lookup('prototype', id);
-          let ownProperties = this.getObjectProperties_(og, id);
-          let prototypePropertyProps = [];
-          // If function exists as a first level interface, use its property
-          // name under window object as interface name, otherwise, use function
-          // name. For example, window.mediaStream and window.webkitMediaStream
-          // are same object with same function name. We do not want to lose
-          // any of the two interfaces.
-          let interfaceName = firstLevel ? name : og.getFunctionName(id);
-          if (prototypePropertyId) {
-            prototypePropertyProps = this.getObjectProperties_(
-                og, prototypePropertyId);
-          }
-          this.arrayMerge_(ownProperties, prototypePropertyProps);
-          let meaningfulProps = this.setMinus_(ownProperties,
-                                               this.builtInFunctionProperties);
-          if (meaningfulProps.length > 0) {
-            // If this function object contains meaningful properties,
-            // add them to our notion of interface members.
-            if (!map[interfaceName]) map[interfaceName] = [];
-            this.arrayMerge_(map[interfaceName],
-                             this.getObjectProperties_(og, id),
-                             this.getObjectProperties_(og, prototypePropertyId));
+      code: function(og) {
+        this.function_ = og.lookup('Function');
+        this.objectDotPrototype_ = og.lookup('Object.prototype');
+        this.functionDotPrototype_ = og.lookup('Function.prototype');
+
+        // Gather functions.
+        const allOgIds = og.getAllIds();
+        for (const id of allOgIds) {
+          if (this.isFunctionLike_(og, id)) {
+            const proto = og.lookup('prototype', id);
+            const names = this.getFunctionNames_(og, id);
+            let protos = this.protos_[proto] =
+                  this.protos_.hasOwnProperty(proto) ? this.protos_[proto] : [];
+            protos.push(id);
+            this.fns_[id] = [];
+            for (const name of names) {
+              this.fns_[id].push(name);
+            }
           }
-          if (og.getPrototype(id)) {
-            this.extractAPI_(og, map, og.getPrototype(id),
-                             `${name}.__proto__`, visited, {proto: true});
+        }
+
+        // Gather "global libraries".
+        const root = og.getRoot();
+        const globalNames = og.getObjectKeys(root);
+        for (const name of globalNames) {
+          const id = og.lookup(name);
+          if ((!og.isType(id)) && (!this.isFunctionLike_(og, id))) {
```

Comment:
I think there's probably a better way of expressing that the first loop should operate on some subset of the input, and that this loop should operate on the remainder. As it is now this check needs to be the negation of the check in the first, which may lead to bugs if it ever needs to be updated. Maybe a single loop that splits the input into 'things that should be processed like functions', 'things that should be processed as library objects', and 'other things', followed by loops over those subsets?